File: DataStructure/53. Maximum Subarray.py
# Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

nums = [-2,1,-3,4,-1,2,1,-5,4]

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        curr_sum, max_sum = nums[0], nums[0]
        for num in nums[1:]:
            if num > num + curr_sum:
                curr_sum = num
            else:
                curr_sum += num
            max_sum = max(curr_sum, max_sum)
        return max_sum

# Using DP
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        DP = nums
        for i in range(1, len(nums)):
            DP[i] = max(DP[i-1] + nums[i], DP[i])
        print(DP)
        return max(DP)

File: LeetCode/DataStructure/53. Maximum Subarray.cs
// Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

int[] nums = new int[] {-2,1,-3,4,-1,2,1,-5,4};


public class Solution {
    public int MaxSubArray(int[] nums) {
        int curr_sum = nums[0];
        int max_sum = nums[0];
        
        for (int i = 1; i < nums.Length; i++)
        {
            if (nums[i] > nums[i] + curr_sum)
            {
                curr_sum = nums[i];
            }
            else 
            {
                curr_sum += nums[i];
            }
            
            if (max_sum < curr_sum)
            {
                max_sum = curr_sum;
            }
            
        }
        
        return max_sum;
    }
}